perfkitbenchmarker/providers/gcp/gcp_dpb_dataflow.py

```diff
@@ -152,4 +186,303 @@ def GetMetadata(self):
+      f'AND resource.labels.project_id = "{self.project}" '
+      f'AND metric.labels.job_id = "{self.job_id}" '
+      f'AND metric.labels.ptransform = "{ptransform}" '
+    )
+
+    aggregation = Aggregation(
+        {
+            "alignment_period": {"seconds": 60},  # 1 minute
+            "per_series_aligner": Aggregation.Aligner.ALIGN_RATE,
+            # "group_by_fields": ["metric.job_id", "metric.ptransform"],
+        }
+    )
+
+    results = client.list_time_series(
+      request={
+        "name": project_name,
+        "filter": api_filter,
+        "interval": interval,
+        "view": monitoring_v3.ListTimeSeriesRequest.TimeSeriesView.FULL,
+        "aggregation": aggregation
+      }
+    )
+
+    if not results:
+      logging.warn('No monitoring data found. Unable to calculate max throughput.')
+      return None
+
+    return self._GetMaxValueFromTimeSeries(results)
+
+  def GetSubscriptionBacklogSize(self, subscription_name, interval_length=4):
+    client = monitoring_v3.MetricServiceClient()
+    project_name = f"projects/{self.project}"
+
+    now = time.time()
+    seconds = int(now)
+
+    interval = TimeInterval(
+        {
+            "end_time": {"seconds": seconds},
+            "start_time": {"seconds": seconds - interval_length * 60},
+        }
+    )
+
+    api_filter = (
+      'metric.type = "pubsub.googleapis.com/subscription/num_undelivered_messages" '
+      'AND resource.labels.subscription_id = "' + subscription_name + '" '
+    )
+    
+    results = client.list_time_series(
+      request={
+        "name": project_name,
+        "filter": api_filter,
+        "interval": interval,
+        "view": monitoring_v3.ListTimeSeriesRequest.TimeSeriesView.FULL,
+      }
+    )
+
+    return self._GetLastValueFromTimeSeries(results)
+
+  def _GetLastValueFromTimeSeries(self, time_series):
+    value = None
+    for i, time_interval in enumerate(time_series):
+      if i != 0: break
+      for j, snapshot in enumerate(time_interval.points):
+        if j != 0: break
+        value = snapshot.value.int64_value
+
+    return value
+
+  def _GetAvgValueFromTimeSeries(self, time_series):
+    """Parses time series data and returns average across intervals.
+
+    Args:
+      time_series: time series of cpu fractional utilization returned by monitoring.
+
+    Returns:
+      Average value across intervals
+    """
+    points = []
+    for time_interval in time_series:
+      for snapshot in time_interval.points:
+        points.append(snapshot.value.double_value)
+
+    if points:
+      # Average over all minute intervals captured
+      averaged = sum(points) / len(points)
+      # If metric unit is a fractional number between 0 and 1 (e.g. CPU utilization metric)
+      # multiply by 100 to display a percentage usage.
+      if time_series.unit == "10^2.%":
+        averaged = round(averaged * 100, 2)
+      return averaged
+    
+    return None 
+
+  def _GetMaxValueFromTimeSeries(self, time_series):
```

Comment:
Does this actually get used in this class? If not I suggest moving it to the subclass that will be in the subsequent PR.

(Noticed this is only used in GetMaxOutputThroughput, which is unused in this PR).